Plugins/Public/base_plugin/FactoryModule.cpp

```diff
@@ -90,181 +125,196 @@ wstring FactoryModule::GetInfo(bool xml)
 bool FactoryModule::Timer(uint time)
 {
 
-	if ((time%set_tick_time) != 0)
+	if ((time % set_tick_time) != 0)
+	{
 		return false;
+	}
 
 	// Get the next item to make from the build queue.
-	if (!active_recipe.nickname && build_queue.size())
+	if (!active_recipe.nickname && !build_queue.empty())
 	{
-		map<uint, RECIPE>::iterator i = recipes.find(build_queue.front());
-		if (i != recipes.end())
-		{
-			active_recipe = i->second;
-		}
+		SetActiveRecipe(build_queue.front());
 		build_queue.pop_front();
 	}
 
 	// Nothing to do.
-	if (!active_recipe.nickname)
-		return false;
-
-	if (Paused)
+	if (!active_recipe.nickname || !base->isCrewSupplied || Paused)
+	{
 		return false;
+	}
 
 	// Consume goods at the cooking rate.
 	bool cooked = true;
-	for (map<uint, uint>::iterator i = active_recipe.consumed_items.begin();
-		i != active_recipe.consumed_items.end(); ++i)
+
+	for (const auto& catalyst : active_recipe.catalyst_items)
+	{
+		uint good = catalyst.first;
+		uint quantityNeeded = catalyst.second;
+
+		uint presentAmount = base->HasMarketItem(good);
+		if ((presentAmount - base->reservedCatalystMap[good]) < quantityNeeded)
+		{
+			return false;
+		}
+		base->reservedCatalystMap[good] += quantityNeeded;
+	}
+
+	if (active_recipe.credit_cost)
 	{
-		uint good = i->first;
-		uint quantity = i->second > active_recipe.cooking_rate ? active_recipe.cooking_rate : i->second;
-		if (quantity)
+		uint moneyToRemove = min(active_recipe.cooking_rate * 10, active_recipe.credit_cost);
```

Comment:
what is this min against `active_recipe.cooking_rate * 10` about?

Reply:
We're consuming credits at 10x the cooking rate of everything else, otherwise it'd be painfully slow if we put in any reasonably large credit amount.

As for min, see above, we don't want to consume more than the recipe actually needs.